File: crafting/julia/Scanner.jl
module Scanners

export Scanner, scantokens

using ..Errors
using ..Lox
using ..Tokens


mutable struct Scanner{S<:AbstractString}
    runner::Runner
    source::S
    tokens::Vector{Any}
    start::UInt
    current::UInt
    line::UInt
end


Scanner(runner, source) = Scanner(
    runner, source, [], one(UInt), one(UInt), one(UInt))


nextchar(scanner::Scanner) = (
    scanner.current = nextind(scanner.source, scanner.current))


const keywords = Dict(
    "and" => token_and,
    "class" => token_class,
    "else" => token_else,
    "false" => token_false,
    "for" => token_for,
    "fun" => token_fun,
    "if" => token_if,
    "nil" => token_nil,
    "or" => token_or,
    "print" => token_print,
    "return" => token_return,
    "super" => token_super,
    "this" => token_this,
    "true" => token_true,
    "var" => token_var,
    "while" => token_while,)


function scantokens(scanner::Scanner)
    while !isatend(scanner)
        scanner.start = scanner.current
        scantokens(scanner)
    end
    push!(scanner.tokens, Token(token_eof, Symbol(""), nothing, scanner.line))
end


isatend(scanner::Scanner) = scanner.current > ncodeunits(scanner.source)


function scantoken(scanner::Scanner)
    char = advance(scanner)
    if char == '('
        addtoken(scanner, token_leftparen)
    elseif char == ')'
        addtoken(scanner, token_rightparen)
    elseif char == '{'
        addtoken(scanner, token_leftbrace)
    elseif char == '}'
        addtoken(scanner, token_rightbrace)
    elseif char == ','
        addtoken(scanner, token_comma)
    elseif char == '.'
        addtoken(scanner, token_dot)
    elseif char == '-'
        addtoken(scanner, token_minus)
    elseif char == '+'
        addtoken(scanner, token_plus)
    elseif char == ';'
        addtoken(scanner, token_semicolon)
    elseif char == '*'
        addtoken(scanner, token_star)
    elseif char == '!'
        addtoken(scanner, match(scanner, '=') ? token_bangequal : token_bang)
    elseif char == '='
        addtoken(
            scanner, match(scanner, '=') ? token_equalequal: token_equal)
    elseif char == '<'
        addtoken(scanner, match(scanner, '=') ? token_lessequal: token_less)
    elseif char == '>'
        addtoken(
            scanner, match(scanner, '=') ? token_greaterequal: token_greater)
    elseif char == '/'
        if match(scanner, '/')
            while peek(scanner) != '\n' && !isatend(scanner)
                advance(scanner)
            end
        else
            addtoken(scanner, token_slash)
        end
    else
        loxerror(scanner.runner, scanner.line, "Unexpected character.")
    end
end


function advance(scanner::Scanner)
    char = scanner.source[scanner.current]
    nextchar(scanner)
    char
end


function addtoken(scanner::Scanner, type::TokenType, literal=nothing)
    endind = prevind(scanner.source, scanner.current)
    lexeme = Symbol(scanner.source[scanner.start:endind])
    push!(scanner.tokens, Token(type, lexeme, literal, scanner.line))
end


function match(scanner::Scanner, expected::AbstractChar)
    isatend(scanner) && return false
    scanner.source[scanner.current] != expected && return false
    nextchar(scanner)
    true
end


isatend(scanner::Scanner) = scanner.current > ncodeunits(scanner.source)


function peek(scanner::Scanner)
    isatend(scanner) ? '\0' : scanner.source[scanner.current]
end


end  # module

File: crafting/python/src/scanner.py
from tokens import Token


class Scanner:
    def __init__(self, source: str):
        self.source = source
        self.tokens: list[Token] = []
        self.start = 0
        self.current = 0
        self.line = 1

    def scan_tokens(self) -> list[Token]:
        while not self.is_at_end():
            self.start = self.current
            self.scan_token()
        self.tokens.apend(Token(TokenType.EOF, '', None, self.line))
        return self.tokens

    def is_at_end(self) -> bool:
        return self.current >= len(self.source)

    def scan_token():
        char = self.advance()
        match char:
            case '(':
                self.add_token(TokenType.LEFT_PAREN)
            case ')':
                self.add_token(TokenType.RIGHT_PAREN)
            case '{':
                self.add_token(TokenType.LEFT_BRACE)
            case '{':
                self.add_token(TokenType.RIGHT_BRACE)
            case ',':
                self.add_token(TokenType.COMMA)
            case '.':
                self.add_token(TokenType.DOT)
            case '-':
                self.add_token(TokenType.MINUS)
            case '+':
                self.add_token(TokenType.PLUS)
            case ';':
                self.add_token(TokenType.SEMICOLON)
            case '*':
                self.add_token(TokenType.STAR)
            case '!':
                (self.add_token(TokenType.BANG_EQUAL) if self.match('=')
                 else self.add_token(TokenType.BANG))
            case '=':
                (self.add_token(TokenType.EQUAL_EQUAL) if self.match('=')
                 else self.add_token(TokenType.EQUAL))
            case '<':
                (self.add_token(TokenType.LESS_EQUAL) if self.match('=')
                 else self.add_token(TokenType.LESS))
            case '>':
                (self.add_token(TokenType.GREATER_EQUAL) if self.match('=')
                 else self.add_token(TokenType.GREATER))
            case '/':
                if self.match('/'):
                    while self.peek() != '\n' and not self.is_at_end():
                        self.advance()
                else:
                    self.add_token(TokenType.SLASH)
            # case ' ':
            #     pass
            # case '\r':
            #     pass
            # case '\t':
            #     pass
            # case '\n':
            #     self.line += 1
            # case '"' | "'":
            #     self._string()
            case _:
            #     if char.isdigit():
            #         self._number()
            #     elif char.isalpha():
            #         self.identifier()
            #     else:
                ErrorHandler.error(self.line, 'Unexpected charcter')
                    
    def advance(self) -> str:
        self.current += 1
        return self.source[self.current - 1]

    def add_token(self, type: TokenType, literal: object = None) -> None:
        text = self.source[self.start:self.current]
        self.tokens.append(Token(type, text, literal, self.line))

    def match(self, exected: str) -> bool:
        if self.is_at_end():
            return False
        if self.source[self.current] != expected:
            return False
        self.current += 1
        return True

    def peek(self) -> str:
        if self.is_at_end():
            return '\0'
        return self.source[self.current]

    def _string():
        pass

    def _number():
        pass

    def identifier():
        
